modules/apps/friendly-url/friendly-url-service/src/main/java/com/liferay/friendly/url/model/impl/FriendlyURLEntryImpl.java

```diff
@@ -5,22 +5,65 @@
 
 package com.liferay.friendly.url.model.impl;
 
+import com.liferay.asset.kernel.model.AssetCategory;
+import com.liferay.asset.kernel.model.AssetEntry;
+import com.liferay.asset.kernel.service.AssetEntryLocalServiceUtil;
 import com.liferay.friendly.url.model.FriendlyURLEntry;
+import com.liferay.friendly.url.model.FriendlyURLEntryLocalization;
 import com.liferay.friendly.url.service.FriendlyURLEntryLocalServiceUtil;
+import com.liferay.petra.string.StringBundler;
 import com.liferay.petra.string.StringPool;
 import com.liferay.portal.kernel.exception.PortalException;
 import com.liferay.portal.kernel.util.LocalizationUtil;
 import com.liferay.portal.kernel.util.Validator;
 
 import java.util.Collection;
 import java.util.Iterator;
+import java.util.List;
 import java.util.Map;
 
 /**
  * @author Pavel Savinov
+ * @author Roberto Díaz
  */
 public class FriendlyURLEntryImpl extends FriendlyURLEntryBaseImpl {
 
+	@Override
+	public String getCategorizedUrlTitle(String languageId) {
+		FriendlyURLEntryLocalization friendlyURLEntryLocalization =
+			FriendlyURLEntryLocalServiceUtil.fetchFriendlyURLEntryLocalization(
+				getPrimaryKey(), languageId);
+
+		if (friendlyURLEntryLocalization == null) {
+			return "";
+		}
+
+		String urlTitle = friendlyURLEntryLocalization.getUrlTitle();
+
+		AssetEntry assetEntry = AssetEntryLocalServiceUtil.fetchEntry(
+			FriendlyURLEntry.class.getName(), getFriendlyURLEntryId());
+
+		if (assetEntry == null) {
+			return urlTitle;
+		}
+
+		List<AssetCategory> categories = assetEntry.getCategories();
+
+		if (categories.isEmpty()) {
+			return urlTitle;
+		}
+
+		String categoryPreffix = "";
+
+		for (AssetCategory category : categories) {
+			categoryPreffix = StringBundler.concat(
+				categoryPreffix, category.getTitle(languageId),
+				StringPool.SLASH);
+		}
+
+		return categoryPreffix + urlTitle;
```

Comment:
You have to use StringBundler to accumulate the result; immediatelly concatenating the string on each iteration defeats the point of using StringBundler. This should be something like this:
```
StringBundler sb = ...;

for (...) {
    sb.append(categoryPrefix);
    sb....
    ...
}

sb.append(urlTitle);

return sb.toString();
```